File: src/camera.ts
import { writeColor } from "./color";
import { HitRecord, Hittable } from "./hittable";
import { interval } from "./interval";
import { ray, Ray } from "./ray";
import { ref, Ref } from "./ref";
import { degreesToRadians, randomNum } from "./utils";
import { color, point3, Vec3, vec3 } from "./vec3";

const sampleSquare = () => vec3(randomNum() - 0.5, randomNum() - 0.5, 0);

export class Camera {
  height: number;
  pixel00Loc!: Vec3;
  pixelDeltaU!: Vec3;
  pixelDeltaV!: Vec3;
  cameraCenter!: Vec3;
  pixelSamplesScale: number;
  canvas: HTMLCanvasElement;
  ctx: CanvasRenderingContext2D;
  width: number;
  aspectRatio: number;
  samplesPerPixel: number;
  maxDepth: number;
  vfov: number;
  lookFrom: Vec3;
  lookAt: Vec3;
  vUp: Vec3;
  defocusAngle: number;
  focusDist: number;
  defocusDiskU!: Vec3;
  defocusDiskV!: Vec3;

  constructor({
    width = 800,
    aspectRatio = 16 / 9,
    samplesPerPixel = 10,
    maxDepth = 10,
    vfov = 90,
    lookFrom = point3(0, 0, 0),
    lookAt = point3(0, 0, -1),
    vUp = vec3(0, 1, 0),
    defocusAngle = 0,
    focusDist = 10,
  }: {
    width?: number;
    aspectRatio?: number;
    samplesPerPixel?: number;
    maxDepth?: number;
    vfov?: number;
    lookFrom?: Vec3;
    lookAt?: Vec3;
    vUp?: Vec3;
    defocusAngle?: number;
    focusDist?: number;
  }) {
    const canvas = document.querySelector("canvas");
    if (!canvas) throw new Error("No canvas");
    this.canvas = canvas;
    const ctx = this.canvas?.getContext("2d");
    if (!ctx) throw new Error("No ctx");
    this.ctx = ctx;

    this.width = width;
    this.aspectRatio = aspectRatio;
    this.samplesPerPixel = samplesPerPixel;
    this.maxDepth = maxDepth;

    // Image
    this.height = Math.floor(width / aspectRatio);
    this.canvas.width = width;
    this.canvas.height = this.height;

    this.pixelSamplesScale = 1 / samplesPerPixel;

    this.vfov = vfov;
    this.lookFrom = lookFrom;
    this.lookAt = lookAt;
    this.vUp = vUp;
    this.defocusAngle = defocusAngle;
    this.focusDist = focusDist;

    this.initialize();
  }

  initialize() {
    this.cameraCenter = this.lookFrom.clone();

    // Camera
    const theta = degreesToRadians(this.vfov);
    const h = Math.tan(theta / 2);
    const viewportHeight = 2.0 * h * this.focusDist;
    const viewportWidth = viewportHeight * (this.width / this.height);

    // Calculate camera basis vectors for camera coordinates
    const w = this.lookFrom.sub(this.lookAt).unit;
    const u = Vec3.cross(this.vUp, w).unit;
    const v = Vec3.cross(w, u);

    const viewportU = u.k(viewportWidth);
    const viewportV = v.k(-viewportHeight);

    this.pixelDeltaU = viewportU.div(this.width);
    this.pixelDeltaV = viewportV.div(this.height);

    const viewportUpperLeft = this.cameraCenter
      .sub(w.k(this.focusDist))
      .sub(viewportU.div(2))
      .sub(viewportV.div(2));
    this.pixel00Loc = viewportUpperLeft.add(
      this.pixelDeltaU.add(this.pixelDeltaV).div(2),
    );

    // camera defocus disk basis vectors
    const defocusRadius =
      this.focusDist * Math.tan(degreesToRadians(this.defocusAngle / 2));
    this.defocusDiskU = u.k(defocusRadius);
    this.defocusDiskV = v.k(defocusRadius);
  }

  render(world: Hittable) {
    for (let j = 0; j <= this.height - 1; j++) {
      console.log(`Rendering scanline ${j}`);
      for (let i = 0; i <= this.width - 1; i++) {
        let pixelColor = vec3(0, 0, 0);

        for (let sample = 0; sample < this.samplesPerPixel; sample++) {
          const r = this.getRay(i, j);
          pixelColor = pixelColor.add(this.rayColor(r, this.maxDepth, world));
        }

        writeColor(this.ctx, pixelColor.k(this.pixelSamplesScale), j, i);
      }
    }
    console.log("Done!");
  }

  rayColor(r: Ray, depth: number, world: Hittable): Vec3 {
    if (depth <= 0) return color(0, 0, 0);

    const rec = new HitRecord();

    const [hasHit, resultRec] = world.hit(r, interval(0.0001, Infinity), rec);
    if (hasHit) {
      let scattered: Ref<Ray> = {};
      let attenuation: Ref<Vec3> = {};

      if (
        resultRec.material!.scatter(ref(r), resultRec, attenuation, scattered)
      ) {
        return this.rayColor(scattered.value!, depth - 1, world).vectorMultiply(
          attenuation.value!,
        );
      } else {
        return color(0, 0, 0);
      }
    }

    const unitDirection = r.direction.unit;
    const a = 0.5 * (unitDirection.y + 1.0);

    return color(1.0, 1.0, 1.0)
      .k(1 - a)
      .add(color(0.5, 0.7, 1.0).k(a));
  }

  getRay(i: number, j: number) {
    const offset = sampleSquare();

    const pixelSample = this.pixel00Loc
      .add(this.pixelDeltaU.k(i + offset.x))
      .add(this.pixelDeltaV.k(j + offset.y));

    const rayOrigin =
      this.defocusAngle <= 0 ? this.cameraCenter : this.defocusDiskSample();
    const rayDirection = pixelSample.sub(rayOrigin);

    return ray(rayOrigin, rayDirection);
  }

  defocusDiskSample() {
    const p = Vec3.randomInUnitDisk();
    return this.cameraCenter
      .add(this.defocusDiskU.k(p.x))
      .add(this.defocusDiskV.k(p.y));
  }
}

File: raytracer/camera.go
package raytracer

import (
	"fmt"
	"image"
	"image/color"
	"math"
	"math/rand/v2"
	"raytracer/math3"
	"runtime"
)

type Camera struct {
	Width            int
	Height           int
	AspectRatio      float64
	SamplesPerPixel  int
	MaxDepth         int
	VFov             float64
	PixelSampleScale float64
	DefocusAngle     float64
	FocusDist        float64
	Center           math3.Vec3
	LookFrom         math3.Vec3
	LookAt           math3.Vec3
	VUp              math3.Vec3
	PixelDeltaU      math3.Vec3
	PixelDeltaV      math3.Vec3
	Pixel00Loc       math3.Vec3
	DefocusDiskU     math3.Vec3
	DefocusDiskV     math3.Vec3
}

type CameraParams struct {
	Width           int
	AspectRatio     float64
	SamplesPerPixel int
	MaxDepth        int
	VFov            float64
	LookFrom        math3.Vec3
	LookAt          math3.Vec3
	DefocusAngle    float64
	FocusDist       float64
}

func NewCamera(params CameraParams) *Camera {
	cam := &Camera{
		Width:            params.Width,
		Height:           int(math.Floor(float64(params.Width) / params.AspectRatio)),
		AspectRatio:      params.AspectRatio,
		SamplesPerPixel:  params.SamplesPerPixel,
		MaxDepth:         params.MaxDepth,
		PixelSampleScale: 1 / float64(params.SamplesPerPixel),
		VFov:             params.VFov,
		DefocusAngle:     params.DefocusAngle,
		FocusDist:        params.FocusDist,
		LookFrom:         params.LookFrom,
		LookAt:           params.LookAt,
		VUp:              math3.Vec3{0, 1, 0},
	}

	cam.Center = cam.LookFrom
	theta := math3.Deg2Rad(cam.VFov)
	h := math.Tan(theta / 2.0)
	viewportHeight := 2 * h * cam.FocusDist
	viewportWidth := viewportHeight * (float64(cam.Width) / float64(cam.Height))
	w := cam.LookFrom.Sub(cam.LookAt).Normalize()
	u := math3.Cross(cam.VUp, w).Normalize()
	v := math3.Cross(w, u)
	viewportU := u.Scale(viewportWidth)
	viewportV := v.Scale(-viewportHeight)
	cam.PixelDeltaU = viewportU.Div(float64(cam.Width))
	cam.PixelDeltaV = viewportV.Div(float64(cam.Height))
	viewportUpperLeft := cam.Center.Sub(w.Scale(cam.FocusDist)).Sub(viewportU.Div(2)).Sub(viewportV.Div(2))
	cam.Pixel00Loc = viewportUpperLeft.Add(cam.PixelDeltaU.Add(cam.PixelDeltaV).Div(2))
	defocusRadius := cam.FocusDist * math.Tan(math3.Deg2Rad(cam.DefocusAngle/2.0))
	cam.DefocusDiskU = u.Scale(defocusRadius)
	cam.DefocusDiskV = v.Scale(defocusRadius)
	return cam
}

func (cam *Camera) Render(world *World, usePool bool) *image.RGBA {
	if !usePool {
		img := image.NewRGBA(image.Rect(0, 0, cam.Width, cam.Height))
		for y := 0; y < cam.Height; y++ {
			fmt.Printf("Rendering scanline %d\n", y)
			for x := 0; x < cam.Width; x++ {
				img.Set(x, y, cam.RenderPixel(x, y, world))
			}
		}
		return img
	}
	return cam.RenderAsync(world)
}

func (cam *Camera) RenderAsync(world *World) *image.RGBA {
	img := image.NewRGBA(image.Rect(0, 0, cam.Width, cam.Height))
	numWorkers := runtime.NumCPU() - 1
	chunkSize := 32
	wp := NewWorkerPool(numWorkers, world)
	totalChunks := int(math.Ceil(float64(cam.Width)/float64(chunkSize))) * int(math.Ceil(float64(cam.Height)/float64(chunkSize)))
	chunks := make([]WorkerJob, 0, totalChunks)
	chunkID := 0

	for y := 0; y < cam.Height; y += chunkSize {
		for x := 0; x < cam.Width; x += chunkSize {
			chunks = append(chunks, WorkerJob{
				XStart: x,
				YStart: y,
				XEnd:   min(x+chunkSize, cam.Width),
				YEnd:   min(y+chunkSize, cam.Height),
				Chunk:  chunkID,
			})
			chunkID++
		}
	}

	rand.Shuffle(len(chunks), func(i, j int) {
		chunks[i], chunks[j] = chunks[j], chunks[i]
	})

	wp.Start(chunkID, img, cam.RenderPixel)
	fmt.Printf("Total jobs: %d\n", chunkID+1)
	for _, job := range chunks {
		wp.Jobs <- job
	}

	wp.Wait()
	return img
}

func (cam *Camera) RenderPixel(x int, y int, world *World) color.Color {
	pixelColor := math3.Vec3{}
	for sample := 0; sample < cam.SamplesPerPixel; sample++ {
		r := cam.GetRay(x, y)
		pixelColor = pixelColor.Add(cam.RayColor(r, cam.MaxDepth, world))
	}
	return convertPixel(pixelColor.Scale(cam.PixelSampleScale))
}

func (cam *Camera) GetRay(x, y int) math3.Ray {
	offsetX, offsetY := rand.Float64()-0.5, rand.Float64()-0.5
	pixelSample := cam.Pixel00Loc.Add(cam.PixelDeltaU.Scale(float64(x) + offsetX)).Add(cam.PixelDeltaV.Scale(float64(y) + offsetY))
	rayOrigin := cam.DefocusDiskSample()
	if cam.DefocusAngle <= 0 {
		rayOrigin = cam.Center
	}
	rayDirection := pixelSample.Sub(rayOrigin)
	return math3.Ray{Origin: rayOrigin, Direction: rayDirection}
}

func (cam *Camera) DefocusDiskSample() math3.Vec3 {
	p := math3.RandomInUnitDisk()
	return cam.Center.Add(cam.DefocusDiskU.Scale(p.X())).Add(cam.DefocusDiskV.Scale(p.Y()))
}

func (cam *Camera) RayColor(r math3.Ray, depth int, world *World) math3.Vec3 {
	if depth <= 0 {
		return math3.Vec3{0.0, 0.0, 0.0}
	}
	if result, hasHit := world.Hit(r, Interval{Min: 0.001, Max: math.MaxFloat64}); hasHit {
		if attenuation, scattered, ok := result.Material.Scatter(r, result); ok {
			survivalScale, shouldTerminate := cam.ShouldTerminateRay(&attenuation, depth)
			if shouldTerminate {
				return math3.Vec3{0.0, 0.0, 0.0}
			}
			if survivalScale > 0 {
				attenuation = attenuation.Scale(1 / survivalScale)
			}
			return cam.RayColor(scattered, depth-1, world).Multiply(attenuation)
		}
		return math3.Vec3{}
	}
	d := r.Direction.Normalize()
	a := 0.5 * (d.Y() + 1.0)
	return math3.Vec3{1.0, 1.0, 1.0}.Scale(1.0 - a).Add(math3.Vec3{0.5, 0.7, 1.0}.Scale(a))
}

func (cam *Camera) ShouldTerminateRay(attenuation *math3.Vec3, depth int) (float64, bool) {
	energy := attenuation.MaxComponent()
	var survivalProb float64
	// Start using Russian Roulette after a few bounces
	if depth < cam.MaxDepth-2 {
		terminationProb := math.Max(0.0, 1.0-energy)
		if rand.Float64() < terminationProb {
			return 0, true
		}

		survivalProb = 1.0 - terminationProb
	}
	return survivalProb, false
}
